text/0000-dyn-trait-declarations.md

```diff
@@ -0,0 +1,123 @@
+
+And finally, this proposal has been proposed to a certain degree before [here](https://github.com/rust-lang/rust/issues/57893#issuecomment-546972824). This RFC would be a way to gradually achieve the goal of this issue comment in the future.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Rust provides support for explicitly marking that a trait must always be object safe. If your trait is being used as an object, you should explicitly mark it as a `dyn` trait as such:
+
+```rust
+dyn trait MyTrait { /* */ }
+```
+
+This will enforce that `MyTrait` is able to be made into a trait object. If this is not the case, then errors in the implementation will be issued. If you were previously using a pattern such as the following, you can simply delete it:
+
+```rust
+fn _assert_is_object_safe(_: &dyn MyTrait) {}
+```
+
+**Note**: note this does not mean traits not marked with `dyn` cannot be made into trait objects. It means the trait is not guaranteed to be object safe and it will not be always automatically enforced.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+When a trait is marked as `dyn`, a further step in type checking is run, which runs the same trait object safety resolution algorithm as the check for a usage such as `&dyn MyTrait`. However, instead of issuing a single error with multiple labels, multiple errors with their own ranges will be issued to accurately reflect the root of the issue.
+
+Parsing this syntax is simple, since `dyn` is a reserved keyword, and there is nothing else that could match in an item context.
+
+The same rules as object safety resolution would be applied, with the exception that every method in the trait must be object safe.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+It is more syntax to think about for the user, it may also introduce confusion if object safety errors are issued twice, once in the usage, and in the implementation block. It may also raise confusion for users who may think traits cannot be made into trait objects without `dyn`.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The pattern of using a trait purely as a trait object or simply enforcing that it must be a trait object is very common, projects such as libcore and rust-analyzer use it in multiple places. This pattern has become common enough that adding a simple language feature to idiomatically support it is viable.
+
+The rationale between the syntax is the following:
+
+- Users already associate `dyn` with trait objects.
+- The syntax is intuitive and unambiguous.
+- The syntax is simple to parse
+
+Another consideration of a proc-macro based approach were brought up too, e.g.
+
+```rust
+#[object_safe]
+trait MyTrait { /* */ }
+```
+
+similar to `#[non_exhaustive]`, this was ultimately ruled out in favor of dedicated syntax.
```

Comment:
Why and where exactly was this ruled out? `#[non_exhaustive]` is a valid attribute and the only way to make `enum`s non-exhaustive currently. So to me it seems like `non_exhaustive` being an attribute would support `#[object_safe]` over dedicated syntax.

Reply:
Sorry, perhaps ruled out was not the right wording. I was referring to a conversation i had on discord, in particular with scottmcm and we decided that it should be dedicated syntax similiar to `dyn Trait`. I am open for using an attribute instead however, although it would need to have accurate ranges in errors, i am not sure if that is possible so perhaps you could clear that up for me? thanks